eng/versioning/pom_file_version_scanner.ps1

```diff
@@ -0,0 +1,438 @@
+                {
+                    return "Error: $($depKey)'s <version> is '$($versionString)' but the dependency version is listed as $($libHash[$depKey].depVer)"
+                }
+            }
+            elseif ($depType -eq $DependencyTypeCurrent) 
+            {
+                if ($versionString -ne $libHash[$depKey].curVer)
+                {
+                    return "Error: $($depKey)'s <version> is '$($versionString)' but the current version is listed as $($libHash[$depKey].curVer)"
+                }
+            }
+            # At this point the version update string, itself, has an incorrect dependency tag
+            else 
+            {
+                return "Error: Invalid dependency type '$($depType)' in version update string $($versionUpdateString). Dependency type must be one of $($DependencyTypeForError)"
+            }
+        }
+    }
+}
+
+# Create one dependency hashtable for libraries we build (the groupIds will make the entries unique) and
+# one hash for external dependencies
+$libHash = @{}
+Build-Dependency-Hash-From-File $libHash $Path\eng\versioning\version_client.txt $false
+Build-Dependency-Hash-From-File $libHash $Path\eng\versioning\version_data.txt $false
+
+$extDepHash = @{}
+Build-Dependency-Hash-From-File $extDepHash $Path\eng\versioning\external_dependencies.txt $true
+
+# Loop through every client and data POM file and perform the verification. Right now
+# management isn't being processed, when it is the checks below will go away and every
+# POM file under the sdk directory will get processed.
+Get-ChildItem -Path $Path -Filter pom*.xml -Recurse -File | ForEach-Object {
+    $pomFile = $_.FullName
+    $xmlPomFile = $null
+
+    if ($_.Name -eq "pom.management.xml")
+    {
+        return
+    }
+
+    if ($PomFilesIgnoreParent -contains $pomFile)
+    {
+        $xmlPomFile = New-Object xml
+        $xmlPomFile.Load($pomFile)
+
+    } else {
+        $xmlPomFile = New-Object xml
+        $xmlPomFile.Load($pomFile)
+        if ($ValidParents -notcontains $xmlPomFile.project.parent.artifactId) 
+        {
+            # This may look odd but ForEach-Object is a cmdlet which means that "continue"
+            # exits the loop altogether and "return" behaves like continue for a particular
+            # loop
+            return
+        }
+    }
+    Write-Output "processing pomFile=$($pomFile)"
+    $dependencyManagement = $xmlPomFile.GetElementsByTagName("dependencyManagement")[0]
+    if ($dependencyManagement)
+    {
+        Write-Output "Error: <dependencyManagement> is not allowed. Every dependency must have its own version and version update tag"
+    }
+
+    # Ensure that the project has a version tag with the exception of projects under the eng directory which
+    # aren't releasing libraries but still need to have their dependencies checked
+    if ($pomFile.Split([IO.Path]::DirectorySeparatorChar) -notcontains "eng") 
```

Comment:
Don't you still want to verify dependencies under eng folder? If the only thing missing from the eng projects is a version then we should probably just add that version and not treat them special.

Reply:
Nevermind I see you are still verifying the other stuff below, but it still might be interesting to add a version to the eng pom files to avoid this special case.

Reply:
@weshaggard The main reason I'd opted to skip the version check for the tools under the eng directory is that we're not shipping them and I didn't see the need to additional bookkeeping to one of the tracking files for them.